pandas/core/arrays/base.py

```diff
@@ -92,7 +92,41 @@ def __getitem__(self, item):
         raise AbstractMethodError(self)
 
     def __setitem__(self, key, value):
-        # type: (Any, Any) -> None
+        # type: (Union[int, np.ndarray], Any) -> None
+        """Set one or more values inplace.
+
+        Parameters
+        ----------
+        key : int or ndarray
+            When called from, e.g. ``Series.__setitem__``, ``key`` will be
+            one of
+
+            * scalar int
+            * ndarray of integers.
+            * boolean ndarray
+            * slice object
+
+        value : ExtensionDtype.type, Sequence[ExtensionDtype.type], or object
+            value or values to be set of ``key``.
+
+        Notes
+        -----
+        This method is not required to satisfy the interface. If an
+        ExtensionArray chooses to implement __setitem__, then some semantics
+        should be observed:
+
+        * Setting multiple values : ExtensionArrays should support setting
+          multiple values at once, ``key`` will be a sequence of integers and
+          ``value`` will be a same-length sequence.
+
+        * Broadcasting : For a sequence ``key`` and a scalar ``value``,
+          each position in ``key`` should be set to ``value``.
+
+        * Coercion : Most users will expect basic coercion to work. For
```

Comment:
this is very hard to support / do. I would not mention this, and its up to the EA subclass to do this (and to be honest kind of breaks the contract, IOW this is supposed to be an integer / slice and not a coercible)

Reply:
It's indeed up to the EA author to decide on this, but this is about coercing `value`, not `key`, so it does not infer with the contract of what `key` can be

Reply:
This section may not be appropriate for the docstring anyway. I'll remove it and add it to the narrative docs (which I've started on another branch).

Reply:
> This section may not be appropriate for the docstring anyway.

To clarify, we have two consumers of the docs. Library authors implementing an EA and users of that library's EA.

To the extent possible, I think we should write docstrings for users of the EA. We can put notes on expected implementation elsewhere.

Reply:
Another possibility (or in addition) is to leave some implementation comments as actual comments. Like

```
def method(..):
    """
    User oriented docstring
    """
    # some additional
    # developer oriented comments
```

as I think it is useful to keep some of those things in the code itself, for someone looking at the base class implementation. 
I think eg `__iter__`, `nbytes`, .. are other examples where we could do such a split